### PR TITLE
Unassign worker from refferal

### PR DESCRIPTION
**Pull request by moarpheus**
## Link to JIRA ticket

[JIRA ticket.](https://hackney.atlassian.net/browse/SCT-1615)

### *What is the problem we're trying to solve*

Previously API for MASH referrals didn't include functionality to un-assign Worker to the Referral. The proposed code changes do that.

### *What changes have we introduced*

Gateway/use case code to support functionality with tests.  
Integration test with + additions in helper to check the full journey.  
Better formatting in gateway tests to improve readability.

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs [hunk 1]
```diff
@@ -41,51 +41,59 @@ public void UpdatingMashReferralThrowsMashReferralNotFoundExceptionWhenGetInfras
         }
 
         [Test]
-        public void UpdatingMashReferralThrowsMashReferralStageMismatchExceptionWhenRequestUpdateIsForScreeningDecisionAndReferralIsNotInScreeningStage()
+        public void
+            UpdatingMashReferralThrowsMashReferralStageMismatchExceptionWhenRequestUpdateIsForScreeningDecisionAndReferralIsNotInScreeningStage()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs:45]: Are we inserting a newline here because of the length of the test name? Is this from running the linter?

Wondering if could use a [description attribute](https://docs.nunit.org/articles/nunit/writing-tests/attributes/description.html) in this case so the the test function name is shorter and we can keep a consistent styling for the tests? 🤔 
e.g
```c#
[Test, Description("Updating Mash Referral Throws Mash Referral Stage Mismatch Exception")]
    public void WhenRequestIsForScreeningDecisionAndReferralIsNotInScreeningStage()
```

File: SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs [hunk 2]
```diff
@@ -185,5 +193,14 @@ public void SuccessfulUpdateOfMashReferralAssignsWorkerUsingEmailAndReturnsMashR
...
         }
 
+        [Test]
+        public void SuccessfulUpdateOfMashReferralUnAssignsWorkerAndReturnsMashReferralDomain()
+        {
+            var mashReferral = MashReferralHelper.SaveMashReferralToDatabase(DatabaseContext);
+            var request = TestHelpers.CreateUpdateMashReferral(updateType: "UNASSIGN-WORKER");
+            var response = _mashReferralGateway.UpdateReferral(request, mashReferral.Id);
+
+            response.Should().BeEquivalentTo(mashReferral.ToDomain());
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs:203]: Is there a way for us to check the value of the worker attribute as part of the assertion?
At a glance, I don't know if the test helper `SaveMashReferralToDatabase` adds a worker to a Mash Referral, could we check this as part of the test and then after the update request check for an expected change to that value?

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        private SocialCareCaseViewerApi.V1.Infrastructure.MashReferral _existingDbReferral;
+
+        [SetUp]
+        public void Setup()
+        {
+            // Clear test database of any rows in the database
+            DatabaseContext.Teams.RemoveRange(DatabaseContext.Teams);
+            DatabaseContext.Workers.RemoveRange(DatabaseContext.Workers);
+            DatabaseContext.MashReferrals.RemoveRange(DatabaseContext.MashReferrals);
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:23]: The test database should be cleared before each test based on the setup method define the IntegrationTestSetup class. Do the tests fail without these lines? We might need to fix the initial setup if that's the case.

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+            // Create existing referral and unrelated worker
+            var existingDbReferral = IntegrationTestHelpers.SaveMashReferralToDatabase(DatabaseContext, "CONTACT");
+            _existingDbReferral = existingDbReferral;
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:27]: You should be able to combine these two lines into one declaration I think 🤔 e.g
```c#
_existingDbReferral = IntegrationTestHelpers.SaveMashReferralToDatabase(DatabaseContext, "CONTACT");
```

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        [Test]
+        public async Task SuccessfulPatchUnAssignsWorkerToReferral()
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:31]: Since this tests a patch, perhaps we could have an assertion that describes what expect before the patch and then what we expect afterward the patch? 

It's a bit hard to follow from the test alone that there's a guaranteed change in the test. For example, if someone changes the test helper code to not add a worker to the referral, then this test wouldn't fail because it's only testing that the worker is null and has not checked that a worker was definitely assigned to the referral.

File: SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs [hunk 6]
```diff
@@ -62,6 +62,28 @@ public void UpdatingMashReferralThrowsWorkerNotFoundExceptionWhenGetWorkerByWork
                 .WithMessage($"Worker with email \"{request.WorkerEmail}\" not found");
         }
 
+        [Test]
+        public void UpdatingMashReferralUnAssignOnSuccessReturnsMashReferralResponse()
+        {
+            var worker = TestHelpers.CreateWorker();
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs:68]: I might have missed it but it doesn't look like we use these variable elsewhere yet. Do we need it?

File: SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs [hunk 7]
```diff
@@ -62,6 +62,28 @@ public void UpdatingMashReferralThrowsWorkerNotFoundExceptionWhenGetWorkerByWork
...
+
+            _mashReferralGateway
+                .Setup(x => x.UpdateReferral(request, referral.Id))
+                .Returns(referral);
+
+            var response = _mashReferralUseCase.UpdateMashReferral(request, referral.Id);
+
+            response.Should().BeEquivalentTo(referral.ToResponse(), options =>
+            {
+                options.Using<DateTime>(ctx => ctx.Subject.Should().BeCloseTo(ctx.Expectation, 1000)).WhenTypeIs<DateTime>();
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs:82]: What do options refer to? Does it affect our response value.

**Review by Renny4Real — COMMENTED** [review]
Looks good 🎉 
Had some questions/comments 🙏🏾

File: SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs (hunk 2 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs:205]: Added explicit set to the WorkerId before the update and checking it equals to null.

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs (hunk 3 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:23]: Copy paste from existing test ( will refactor that too )

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs (hunk 4 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:27]: Copy paste from existing test ( will refactor that too )

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs (hunk 3 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:23]: Checked the IntegrationTestSetup it should wipe before and after, tested and it works indeed.

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs (hunk 5 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:31]: Added an assertion to check that WorkerId is not null before patch.

File: SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs (hunk 6 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs:68]: Nope we don't

File: SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs (hunk 7 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs:82]: In Fluent Assertions when comparing objects with DateTime properties there are sometimes a slight mismatch in the milliseconds and the comparison fail. The way we get around it is to set the comparison option like so.

File: SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs (hunk 1 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs:45]: Up for discussion with rest of the team, I personally think idea is great since it is difficult to read the test names.

Comment (Renny4Real) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/MashReferralGatewayTests/UpdateReferralTests.cs:45]: Yep. I agree with a team discussion 👍🏾

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/AssignWorkerToMashReferralOnContactDecisionTests.cs [hunk 8]
```diff
@@ -19,9 +19,9 @@ public class AssignWorkerToMashReferralOnContactDecisionTests : IntegrationTestS
         public void Setup()
         {
             // Clear test database of any rows in the database
-            DatabaseContext.Teams.RemoveRange(DatabaseContext.Teams);
-            DatabaseContext.Workers.RemoveRange(DatabaseContext.Workers);
-            DatabaseContext.MashReferrals.RemoveRange(DatabaseContext.MashReferrals);
+            // DatabaseContext.Teams.RemoveRange(DatabaseContext.Teams);
+            // DatabaseContext.Workers.RemoveRange(DatabaseContext.Workers);
+            // DatabaseContext.MashReferrals.RemoveRange(DatabaseContext.MashReferrals);
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/AssignWorkerToMashReferralOnContactDecisionTests.cs:24]: Perhaps delete the commented out code.

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        public void Setup()
+        {
+            // Create existing referral and unrelated worker
+            _existingDbReferral = IntegrationTestHelpers.SaveMashReferralToDatabase(DatabaseContext, "CONTACT");
+        }
+
+        [Test]
+        public async Task SuccessfulPatchUnAssignsWorkerFromReferral()
+        {
+            _existingDbReferral.WorkerId.Should().NotBeNull();
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/UnAssignWorkerFromMashReferralTests.cs:27]: Nice! 🙌🏾

File: SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs (hunk 7 above)
Comment (Renny4Real) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/MashReferral/UpdateMashReferralTests.cs:81]: Thanks! 🙌🏾

**Review by Renny4Real — APPROVED** [review]
Looks good! 🎉

File: SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/AssignWorkerToMashReferralOnContactDecisionTests.cs (hunk 8 above)
Comment (moarpheus) [reply SocialCareCaseViewerApi.Tests/V1/IntegrationTests/MASH/AssignWorkerToMashReferralOnContactDecisionTests.cs:24]: Ouch ...
